src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1214,4 +1233,45 @@ protected Engine newEngine() {
     public boolean allowsPrimaryPromotion() {
         return true;
     }
+
+    // pkg private for testing
+    void persistMetadata(ShardRouting newRouting, ShardStateMetaData oldShardStateMetadata) {
```

Comment:
It's currently remote controlled by the creation of oldShardStateMetadata in updateRoutingEntry. If we want to keep the optional persistence based on the previous state,  we can do it directly on currentRouting  have it all in one place (either in updateRoutingEntry, or replacing persistMetadata.  Something like this:

```
 if (persistState) {
                final String writeReason;
                if (newRouting.active() == false) {
                    logger.trace("skipping state persistence as it's not active");
                    writeReason = null;
                } else if (currentRouting == null) {
                    writeReason = "freshly started, version [" + newRouting.version() + "]";
                } else if (currentRouting.version() < newRouting.version()) {
                    writeReason = "version changed from [" + currentRouting.version() + "] to [" + newRouting.version() + "]";
                } else if (currentRouting.equals(newRouting) == false) {
                    writeReason = "routing changed from " + currentRouting + "] to " + newRouting;
                } else {
                    logger.trace("skip writing shard state. previous version:[{}] current version: [{}]",currentRouting.version(), newRouting.version());
                    writeReason = null;
                }
                if (writeReason != null) {
                    final ShardStateMetaData shardStateMetaData = new ShardStateMetaData(newRouting.version(), newRouting.primary(), getIndexUUID());
                    try {
                    ShardStateMetaData.write(logger, writeReason, shardId, shardStateMetaData, currentRouting != null, nodeEnv.shardPaths(shardId));
                    } catch (IOException e) { // this is how we used to handle it.... :(
                        logger.warn("failed to write shard state for shard " + shardId, e);
                        // we failed to write the shard state, we will try and write
                        // it next time...
                    }
                }
            }
```